src/dotty/tools/dotc/transform/Space.scala

```diff
@@ -0,0 +1,316 @@
+      ss.exists(subspace(a, _)) ||
+      (canDecompose(t1) && subspace(Or(getPartitions(t1).map(Typ)), b))
+    case (Typ(t1), Kon(t2, ss)) =>
+      isSubType(t1, t2) && subspace(Kon(t2, getSignature(t2).map(Typ)), b)
+    case (Or(ss), _) =>
+      ss.forall(subspace(_, b))
+    case (Kon(t1, ss), Typ(t2)) =>
+      isSubType(t1, t2) ||
+      simplify(a) == Empty ||
+      (isSubType(t2, t1) &&
+        canDecompose(t1) &&
+        subspace(Or(getPartitions(t1).map(Typ)), b))
+    case (Kon(t1, ss1), Or(ss2)) =>
+      simplify(minus(a, b)) == Empty
+    case (Kon(t1, ss1), Kon(t2, ss2)) =>
+      isEqualType(t1, t2) && ss1.zip(ss2).forall((subspace _).tupled)
+  }
+
+  // intersection of two spaces, non-commutative(inherit case class)
+  def intersect(a: Space, b: Space): Space = (a, b) match {
+    case (Empty, _) | (_, Empty) => Empty
+    case (_, Or(ss)) => Or(ss.map(intersect(a, _)))
+    case (Or(ss), _) => Or(ss.map(intersect(_, b)))
+    case (Typ(t1), Typ(t2)) =>
+      if (isSubType(t1, t2)) a
+      else if (isSubType(t2, t1)) b
+      else Empty
+    case (Typ(t1), Kon(t2, ss)) =>
+      if (isSubType(t2, t1)) b
+      else if (isSubType(t1, t2)) a
+      else Empty
+    case (Kon(t1, ss), Typ(t2)) =>
+      if (isSubType(t1, t2) || isSubType(t2, t1)) a
+      else Empty
+    case (Kon(t1, ss1), Kon(t2, ss2)) =>
+      if (!isEqualType(t1, t2)) Empty
+      else if (ss1.zip(ss2).exists(p => simplify(intersect(p._1, p._2)) == Empty)) Empty
+      else
+        Kon(t1, ss1.zip(ss2).map((intersect _).tupled))
+  }
+
+  //  the space of a not covered by b
+  def minus(a: Space, b: Space): Space = (a, b) match {
+    case (Empty, _) => Empty
+    case (_, Empty) => a
+    case (Typ(t1), Typ(t2)) =>
+      if (isSubType(t1, t2)) Empty
+      else if (isSubType(t2, t1) && canDecompose(t1))
+        minus(Or(getPartitions(t1).map(Typ)), b)
+      else a
+    case (Typ(t1), Kon(t2, ss)) =>
+      if (isSubType(t1, t2)) minus(Kon(t2, getSignature(t2).map(Typ)), b)
+      else if (isSubType(t2, t1) && canDecompose(t1))
+        minus(Or(getPartitions(t1).map(Typ)), b)
+      else a
+    case (_, Or(ss)) =>
+      ss.foldLeft(a)(minus)
+    case (Or(ss), _) =>
+      Or(ss.map(minus(_, b)))
+    case (Kon(t1, ss), Typ(t2)) =>
+      if (isSubType(t1, t2)) Empty
+      else if (simplify(a) == Empty) Empty
+      else if (isSubType(t2, t1) && canDecompose(t1))
+        minus(Or(getPartitions(t1).map(Typ)), b)
+      else a
+    case (Kon(t1, ss1), Kon(t2, ss2)) =>
+      if (!isEqualType(t1, t2)) a
+      else if (ss1.zip(ss2).exists(p => simplify(intersect(p._1, p._2)) == Empty)) a
+      else if (ss1.zip(ss2).forall((subspace _).tupled)) Empty
+      else
+        Or(
+          ss1.zip(ss2).map((minus _).tupled).zip(0 to ss2.length - 1).map {
+            case (ri, i) => Kon(t1, ss1.updated(i, ri))
+          })
+  }
+}
+
+// scala implementation of space logic
+class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
```

Comment:
do `import tpd._` here so that you can use `Tree` instead of `tpd.Tree` below, a lot of phases do this.